module/zfs/btree.c

```diff
@@ -0,0 +1,2191 @@
+	 * right merging node and the separator. This may cause further
+	 * merging.
+	 */
+	btree_hdr_t *new_rm_hdr;
+
+	if (l_hdr != NULL) {
+		ASSERT(l_hdr->bth_core);
+		btree_core_t *left = (btree_core_t *)l_hdr;
+
+		if (btree_verify_intensity >= 5) {
+			for (int i = 0; i < hdr->bth_count + 1; i++) {
+				btree_verify_poison_at(tree, l_hdr,
+				    l_hdr->bth_count + i);
+			}
+		}
+		/* Move the separator into the left node. */
+		uint8_t *e_out = left->btc_elems + l_hdr->bth_count * size;
+		uint8_t *separator = parent->btc_elems + (parent_idx - 1) *
+		    size;
+		bcopy(separator, e_out, size);
+
+		/* Move all our elements and children into the left node. */
+		bt_transfer_core(tree, node, 0, idx - 1, left,
+		    l_hdr->bth_count + 1, BSS_TRAPEZOID);
+		bt_transfer_core(tree, node, idx, hdr->bth_count - idx + 1,
+		    left, l_hdr->bth_count + idx, BSS_PARALLELOGRAM);
+
+		/* Reparent all our children to point to the left node. */
+		btree_hdr_t **new_start = left->btc_children +
+		    l_hdr->bth_count + 1;
+		for (int i = 0; i < hdr->bth_count + 1; i++)
+			new_start[i]->bth_parent = left;
+
+		/* Update bookkeeping */
+		l_hdr->bth_count += hdr->bth_count + 1;
+		for (int i = 0; i <= l_hdr->bth_count; i++) {
+			ASSERT3P(left->btc_children[i]->bth_parent, ==, left);
+			ASSERT3P(left->btc_children[i], !=, rm_hdr);
+		}
+		ASSERT3U(l_hdr->bth_count, ==, BTREE_CORE_ELEMS);
+		new_rm_hdr = hdr;
+	} else {
+		ASSERT3P(r_hdr, !=, NULL);
+		ASSERT(r_hdr->bth_core);
+		btree_core_t *right = (btree_core_t *)r_hdr;
+
+		if (btree_verify_intensity >= 5) {
+			for (int i = 0; i < r_hdr->bth_count; i++) {
+				btree_verify_poison_at(tree, hdr,
+				    hdr->bth_count + i + 1);
+			}
+		}
+		/*
+		 * Overwrite rm_hdr and its separator by moving node's
+		 * elements and children forward.
+		 */
+		bt_shift_core_left(tree, node, idx, hdr->bth_count - idx + 1,
+		    BSS_PARALLELOGRAM);
+
+		/*
+		 * Move the separator to the first open spot in node's
+		 * elements.
+		 */
+		uint8_t *e_out = node->btc_elems + (hdr->bth_count - idx) *
+		    size;
+		uint8_t *separator = parent->btc_elems + parent_idx * size;
+		bcopy(separator, e_out, size);
+
+		/* Move the right node's elements and children to node. */
+		bt_transfer_core(tree, right, 0, r_hdr->bth_count, node,
+		    hdr->bth_count - idx + 1, BSS_TRAPEZOID);
+
+		/* Reparent the right node's children to point to node. */
+		for (int i =  hdr->bth_count - idx; i < r_hdr->bth_count + 1;
+		    i++) {
+			node->btc_children[i]->bth_parent = node;
+		}
+
+		/* Update bookkeeping. */
+		hdr->bth_count += r_hdr->bth_count + 1;
+		for (int i = 0; i <= hdr->bth_count; i++) {
+			ASSERT3P(node->btc_children[i]->bth_parent, ==, node);
+			ASSERT3P(node->btc_children[i], !=, rm_hdr);
+		}
+
+		ASSERT3U(hdr->bth_count, ==, BTREE_CORE_ELEMS);
+		new_rm_hdr = r_hdr;
+	}
+
+	new_rm_hdr->bth_count = 0;
+	btree_node_destroy(tree, new_rm_hdr);
+	btree_remove_from_node(tree, parent, new_rm_hdr);
```

Comment:
Is this relying on tail-call optimization to avoid recursion?  If we need the tail-call optimization to not overflow the stack, is there some way to guarantee that we get it? (some compiler annotation?)  I imagine there's a similar case for insertion as well?

Reply:
Yes, it is, and I don't believe there's any way to guarantee that. I'm not especially worried about overflowing the stack, though. It's very hard to get into a situation where you need more than a few levels of the btree, and almost impossible to get to one that's over 5 levels. A 5 level btree can have (128 + 128^2 + 128^3 + 128^4 + 128^4*(4060/24))/2 = 22 billion entries in the worst case.